pkg/doc/verifiable/credential.go

```diff
@@ -452,90 +436,121 @@ func decodeCredentialSchema(data []byte) ([]TypedID, error) {
 }
 
-// NewCredential creates an instance of Verifiable Credential by reading a JSON document from bytes.
-// It also applies miscellaneous options like custom decoders or settings of schema validation.
-func NewCredential(vcData []byte, opts ...CredentialOpt) (*Credential, error) {
-	// Apply options
-	crOpts := defaultCredentialOpts()
-
-	for _, opt := range opts {
-		opt(crOpts)
-	}
+// NewCredential decodes Verifiable Credential from bytes which could be marshalled JSON or serialized JWT.
+// It also applies miscellaneous options like settings of schema validation.
+// It returns decoded Credential and its marshalled JSON.
+// For JSON bytes input, the output marshalled JSON is the same value.
+// For serialized JWT input, the output is the result of decoding `vc` claim from JWT.
+// The output Credential and marshalled JSON can be used for extensions of the base data model
+// by checking ExtraFields of Credential and/or unmarshalling the JSON to custom date structure.
+func NewCredential(vcData []byte, opts ...CredentialOpt) (*Credential, []byte, error) {
+	// Apply options.
+	crOpts := parseCredentialOpts(opts)
 
+	// Decode credential (e.g. from JWT).
 	vcDataDecoded, err := decodeRaw(vcData, crOpts)
 	if err != nil {
-		return nil, err
+		return nil, nil, fmt.Errorf("decode new credential: %w", err)
 	}
 
-	// unmarshal VC from JSON
+	// Unmarshal raw credential from JSON.
 	var raw rawCredential
-
 	err = json.Unmarshal(vcDataDecoded, &raw)
+
 	if err != nil {
-		return nil, fmt.Errorf("new credential: %w", err)
+		return nil, nil, fmt.Errorf("unmarshal new credential: %w", err)
 	}
 
-	schemas, err := loadCredentialSchemas(&raw, vcDataDecoded)
-	if err != nil {
-		return nil, err
+	// Load custom credential schemas if defined.
+	var schemas []TypedID
+	if raw.Schema != nil {
+		schemas, err = loadCredentialSchemas(vcDataDecoded)
+		if err != nil {
+			return nil, nil, fmt.Errorf("load schemas of new credential: %w", err)
+		}
+	} else {
+		schemas = make([]TypedID, 0)
 	}
 
+	// Validate raw credential.
 	err = validate(vcDataDecoded, schemas, crOpts)
 	if err != nil {
-		return nil, err
+		return nil, nil, fmt.Errorf("validate new credential: %w", err)
 	}
 
-	cred := crOpts.template()
-
-	err = cred.fill(&raw)
+	// Create credential from raw.
+	vc, err := newCredential(&raw, schemas)
 	if err != nil {
-		return nil, err
+		return nil, nil, fmt.Errorf("build new credential: %w", err)
 	}
 
-	cred.Schemas = schemas
+	return vc, vcDataDecoded, nil
+}
 
-	for _, decoder := range crOpts.decoders {
-		err = decoder(vcDataDecoded, cred)
-		if err != nil {
-			return nil, err
+// ExtCredentialProducer is a factory for Credentials with extended data model.
+type ExtCredentialProducer interface {
+	Accept(vc *Credential) bool
+	Apply(vc *Credential, dataJSON []byte) (interface{}, error)
+}
+
+// NewExtCredential creates custom extended credentials from bytes which could be marshalled JSON or serialized JWT.
```

Comment:
This sounds like the same Input Bytes as NewCredential. This is confusing to the end developer. e.g.,

why does NewCredential need to return these bytes? are these byes different somehow?
why have a second new function?

Reply:
I've changed the description of `NewCredential()`, there is the rationale why we return the bytes as second return value. Are you thinking about merging `NewCredential()` and `NewExtCredential` together? The reason I added these two functions is that in the second case we return `interface{}` just because we return the value which `ExtCredentialProducer::Apply()` returns (what is extended VC).

Reply:
- Yes I suspected that this new function was created so that an untyped interface could be returned.
- I wasn't entirely sure about the second return value. (e.g., is it different than the inputs, should it be a field vs a second return value, ...)
  - I was also wondering if `NewExtCredential` should take bytes or a *Credential.
- I also wonder if the name `NewExtCredential` is correct.

Reply:
- If the second return value is only used for extension and/or unmarshalling to custom structures, I think `NewCredential` should never have been called and rather `NewExtCredential` should've been used. I don't see why the business logic should have the option of reinventing the wheel in this scenario.

- @troyronda `NewExtCredential` doesn't return an `ExtCredential` type because it can't have a typed struct encompassing all possibilities. It does, however, return an "extended credential" through the interface. I would however change this to `NewExtendedCredential` instead as more verbose naming, in this case, would be better for readability.

Reply:
- to extend NewCredential(), `decoders` and `templates` were passed previously as `credentialOpts`. The `decoder` defined extra decoding actions. The `template` allowed to define a factory method of `Credential`.
- I decided to simplify that and make the extension of `Credential` an explicit two-step. The first step is to decode base `Credential`. The second one is to make extra decodings if needed. I found useful to return 2 results of decoding at `NewCredential()` - the `Credential` struct and `[]byte`. For extension, it's really easy to unmarshal the rest of the fields (which are missing at base `Credential`) from that `[]byte` rather than check `ExtraFields` map of `Credential`. Note that return `[]byte` is the same as input `[]byte` (of `NewCredential()`) for JSON case, but is a result of parsing and refinement in the case of JWT.
- `NewExtCredential()`  was added as actually the goal of the task. This pattern was added at `credential_extension_switch_test.go` initially, to show the case when we can (e.g. as **Holder** side) received credentials of several different kinds and want to apply appropriate extension based on e.g. `context` and `type`. I agree that we shouldn't use `New...()` name, instead use smth like `Create...()`. I don't know how to name this better, maybe `CreateCustomCredential()` will be more clear for the user.

Reply:
Same thoughts as Haardik, the very minor suggestions. Overall it looks good, and these features will allow us to progress with our OTBS demo using VC's

Reply:
I think we should have 2 public functions `NewCredential()` and `CreateCustomCredential()`.
- `NewCredential()` can be used when we don't want to introduce extra structures for extensions and instead want to have an output of explicit type (`Credential`) and get access to all custom fields at `ExtraFields` map of `Credential`.
- `CreateCustomCredential()` can be used for explicit extensions. A special use case is when  we expect to receive two or more different extension types.